### PR TITLE
OCPBUGS-17113: rework `unstructured` types for monitoring 

### PR DESCRIPTION
**Pull request by rexagod**
This PR incorporates the following changes:

* refactoring the existing functions (and comments) a bit,
* cache-based apply for `unstructured` monitoring types, and,
* adding explicit `*KnownUnstructured` functions for `alertmanager` and `prometheus`.

***

[Slack discussion thread for the same.](https://redhat-internal.slack.com/archives/CB48XQ4KZ/p1694071423825099)

/cc @dgrisonnet

### CODE REVIEW

**Comment by rexagod** [comment]
- [ ] Additionally, look into utilizing `observedGeneration` and `resourceVersion` fields whenever present to better cache based on object `spec` and `metadata` changes.

**Comment by dgrisonnet** [comment]
I completely missed that PR, @rexagod is this still relevant?

**Comment by rexagod** [comment]
Tests seem to be failing since earlier we only checked for `spec` changes when `Apply`ing a resource, unlike now, where we fail-fast if the `metadata` is the same. This functionality cannot be tested in a unit-test currently since we rely on `resourceVersion` and `generation` now, so I'll add an e2e test (which is a bit odd since I don't see those being run in the workflow, given there already exist some of those).

**Comment by rexagod** [comment]
Ah, [makes sense](https://github.com/openshift/release/blob/master/ci-operator/jobs/openshift/library-go/openshift-library-go-master-presubmits.yaml#L18).

**Comment by rexagod** [comment]
I'd prefer a single `test-e2e` target in the aforementioned Prow definition, which would prevent us from having to add a new e2e target every-time it's introduced here, in `/release`.

**Comment by rexagod** [comment]
(friendly ping @dgrisonnet)

**Comment by rexagod** [comment]
(friendly ping @dgrisonnet)

**Comment by rexagod** [comment]
cc @deads2k

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (deads2k) [inline pkg/operator/resource/resourceapply/monitoring.go:87]: `hashOfResourceStruct` is unlikely to work on the map.  Either a unit test showing that it functions properly for `*unstructure.Unstructured` or update to have it hash the kuberentes-sigs yaml serialization.

Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:87]: Added a unit-test.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:127]: ACK, I've moved this below the `spec` check.

**Comment by rexagod** [comment]
(bump)

**Comment by dusk125** [comment]
lgtm

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (deads2k) [inline pkg/operator/resource/resourceapply/monitoring.go:124]: If spec did not change, but labels do change.  How does this code avoid returning early?

Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go]: Addressed in the latest commit, also added some tests.

File: pkg/operator/resource/resourceapply/monitoring_test.go
Comment (rexagod) [inline pkg/operator/resource/resourceapply/monitoring_test.go:207]: Casting actions' verbs to a set here to account for flakes such as the one below (owing to order guarantees).
```

```

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (machine424) [inline pkg/operator/resource/resourceapply/monitoring.go:74]: why do we need a deepCopy here? I can't see any call that may alter it.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (machine424) [inline pkg/operator/resource/resourceapply/monitoring.go:124]: IIUC, the default values set by the server (that are not overwritten) are ignored, thus in most cases, it will consider the objects as not equal.

Do we plan to allow overwriting `noDefaulting` and  `equality.Semantic` in the future?
In the meanwhile I think we should state it explicitly (as "improved" in `XXXImproved` may be misleading): "if you want to make sure UPDATES are only triggered when needed, you should enable the cache."

CMO doesn't use the cache for example (for now), as the no-cache `ApplyXXX` utils it's using now ensure limited UPDATES even without the cache.

**Review by machine424 — COMMENTED** [review]
I suggest we shorten the feedback loop and try this on CMO before the merge/release.
And that could be considered as an extra test.
I can help with that.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:124]: > Do we plan to allow overwriting noDefaulting and  equality.Semantic in the future?

I'm +1 on moving `noDefaulting` and `equality.Semantic` parameters to the function signature.

> In the meanwhile I think we should state it explicitly (as "improved" in XXXImproved may be misleading): "if you want to make sure UPDATES are only triggered when needed, you should enable the cache."

I'm not sure what is being referred to here? Are you suggesting a comment?

> CMO doesn't use the cache for example (for now), as the no-cache ApplyXXX utils it's using now ensure limited UPDATES even without the cache.

I believe we will need to initialize a cache for CMO to be able to benefit from the changes being made here, i.e., dropping `ApplyXXXImproved` in favor of `ApplyUnstructuredResourceImproved`.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:74]: Dropped.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go]: Friendly ping. :)

**Review by dusk125 — CHANGES_REQUESTED** [review]

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:104]: Ah! It seems I made the last change in a hurry, fixed.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go]: Additionally, we might want to supply custom a `equalityChecker` or `defaultingFunc` that caters to the CRD to make things more robust.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (jan--f) [inline pkg/operator/resource/resourceapply/monitoring.go:110]: Should these error returns maybe contain context about the object where the error occurred? Here its the existing object, further down the required?

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (jan--f) [inline pkg/operator/resource/resourceapply/monitoring.go:158]: Should this function not check if `required` has a `ResourceVersion` set and if so if it corresponds to the existing object before updating? Wouldn't this otherwise risk overwriting prior updates?

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:110]: Hmm, so `found` here corresponds to the [`existingObject`](https://github.com/openshift/library-go/pull/1575/files#diff-55de3da9293d86816615a21ec35c9725ab821c04b938de22d815679554b2fe92R105) and the [`requiredObject`](https://github.com/openshift/library-go/pull/1575/files#diff-55de3da9293d86816615a21ec35c9725ab821c04b938de22d815679554b2fe92R112) below. Do you mean appending something more than just the object name in the error messages? LMK if I missed something.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:158]: I believe due to the fact that:

* the resource version cache updates [here](https://github.com/openshift/library-go/blob/e6bb5506390c06b624c7192bc41ed4fc4b517618/pkg/operator/resource/resourceapply/resource_cache.go#L149) are always sourced from the existing state and never the required one, and,
* an explicit required resource version that is not equal to the existing one will cause the Kube API to throw an error,

should ensure a valid and latest copy of the resource version is maintained in the cache, and used for skipping updates. WDYT?

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (jan--f) [reply pkg/operator/resource/resourceapply/monitoring.go:110]: Sure I understand when the error is returned. But iiuc the error returned will always be something like `metadata not found in prometheus-k8s`, whether its the existing copy or the required object. Adding information which object generates an error might be a good idea?

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:158]: Right, the write happens [here](https://github.com/openshift/library-go/blob/e6bb5506390c06b624c7192bc41ed4fc4b517618/pkg/operator/resource/resourceapply/resource_cache.go#L117) where the cache key is dependent on the existing object's `ResourceVersion`. I've added a fail-fast check.

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:110]: Oh, you mean the object key (as in `namespace/object`) for a better context? Done!

**Comment by jan--f** [comment]
Thanks, this
/lgtm

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (rexagod) [reply pkg/operator/resource/resourceapply/monitoring.go:140]: @machine424 suggested we drop these parameters, since we will only be introducing this function for the Prometheus object for now. If there are any defaults that need to be set for it, they should be set in Prometheus operator, which makes sense.

However, when and if we decide to expand this in the future, we will need to add these to cater to other components' use cases that don't have such middlewares (and require setting defaults in CMO itself, as that's where their configuration is defined).

**Comment by rexagod** [comment]
@jan--f Can you please `lgtm` again, I pushed a fixup commit that I dropped after a while.

**Comment by jan--f** [comment]
/lgtm
:+1:

**Comment by rexagod** [comment]
Rebased, needs `/lgtm`.

**Comment by rexagod** [comment]
Friendly ping. 🙂 

cc @deads2k

File: pkg/operator/resource/resourceapply/monitoring.go
Comment (p0lyn0mial) [reply pkg/operator/resource/resourceapply/monitoring.go:166]: Unfortunately, this doesn't seem to work because the server requires an RV to be set when performing an update. This manifests as (the operator is permanently broken):
```
{Failed to upgrade kube-apiserver, operator was degraded (KubeAPIServerStaticResources_SyncError): KubeAPIServerStaticResourcesDegraded: "assets/alerts/api-usage.yaml" (string): prometheusrules.monitoring.coreos.com "api-usage" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update
KubeAPIServerStaticResourcesDegraded:   }
```

The issue is quite serious, as it will block upgrades to `4.18`. Please take a look at https://github.com/openshift/cluster-kube-apiserver-operator/pull/1750, especially the `e2e-aws-ovn-upgrade` job.

@rexagod , are you willing to work on a fix? I think we would like to have a test to prevent future bugs.
If the code is part of 4.17 I think we should back port the fix and the test.

**Comment by p0lyn0mial** [comment]
Unfortunately this PR doesn't work as expected, see https://github.com/openshift/library-go/pull/1575#discussion_r1791701946 for more info.

**Pull request merged**
